### PR TITLE
Initial import

### PR DESCRIPTION
**Pull request by ttddyy**
Here is the import from datasource-proxy-r2dbc.

- Copy and paste from datasource-proxy-r2dbc
- Modified package structure
- Added license header
- Incorporated `Wrapped` interface (in separate commit)

### CODE REVIEW

**Comment by ttddyy** [comment]
Hmm, javadoc generation is failing. Will fix them.

File: src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    ConnectionFactory createConnectionFactory(ConnectionFactory connectionFactory);
+
+    Connection createConnection(Connection connection, ConnectionInfo connectionInfo);
+
+    Batch<?> createBatch(Batch<?> batch, ConnectionInfo connectionInfo);
+
+    Statement<?> createStatement(Statement<?> statement, String query, ConnectionInfo connectionInfo);
+
+    Result createResult(Result result, QueryExecutionInfo executionInfo);
+
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java]: Some of these methods sound strange. `ConnectionFactory createConnectionFactory(ConnectionFactory connectionFactory)` for example. `createConnectionFactory` implies it's creating a new one, but it has one as an input. Does this wrap one with another?

File: src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        }
+        return null;
+    }
+
+    public static Connection getOriginalConnection(Result result) {
+        if (result instanceof ConnectionHolder) {
+            return ((ConnectionHolder) result).getOriginalConnection();
+        }
+        return null;
+    }
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java]: Is this `null` approach the right approach? Does wrapping with `Optional` cost too much?

File: src/main/java/io/r2dbc/proxy/callback/ReactiveBatchCallback.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+ * @author Tadaya Tsuyukubo
+ */
+public class ReactiveBatchCallback extends CallbackSupport {
+
+    private Batch<?> batch;
+
+    private ConnectionInfo connectionInfo;
+
+    private List<String> queries = new ArrayList<>();
+
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/callback/ReactiveBatchCallback.java]: Should this initialization be moved into the constructor call?

File: src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+        Object result = proceedExecution(method, this.connection, args, this.proxyConfig.getListeners(), this.connectionInfo, null, onComplete);
+
+        if ("createBatch".equals(methodName)) {
+            return this.proxyConfig.getProxyFactory().createBatch((Batch) result, this.connectionInfo);
+        } else if ("createStatement".equals(methodName)) {
+            String query = (String) args[0];
+            return this.proxyConfig.getProxyFactory().createStatement((Statement) result, query, this.connectionInfo);
+        }
+
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java]: Is keying off every API's method name via String the only way to manage this?

File: src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionFactoryCallback.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+
+        String methodName = method.getName();
+
+        if ("unwrap".equals(methodName)) {
+            return this.connectionFactory;
+        }
+
+        BiFunction<Object, MethodExecutionInfo, Object> onMap = null;
+        if ("create".equals(methodName)) {
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionFactoryCallback.java]: Line break

File: src/main/java/io/r2dbc/proxy/listener/CompositeProxyExecutionListener.java [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    public List<ProxyExecutionListener> getListeners() {
+        return this.listeners;
+    }
+
+//    public void setListeners(List<ProxyExecutionListener> listeners) {
+//        this.listeners = listeners;
+//    }
+
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/listener/CompositeProxyExecutionListener.java]: Remove.

File: src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+     * @param executionInfo query execution context
+     */
+    default void onMethodExecution(MethodExecutionInfo executionInfo) {
+        ProxyEventType eventType = executionInfo.getProxyEventType();
+        if (eventType == ProxyEventType.BEFORE_METHOD) {
+            beforeMethod(executionInfo);
+        } else if (eventType == ProxyEventType.AFTER_METHOD) {
+            afterMethod(executionInfo);
+        }
+    }
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java]: And what if neither condition is met?

File: src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+     * @param executionInfo query execution context
+     */
+    default void onQueryExecution(QueryExecutionInfo executionInfo) {
+        ProxyEventType eventType = executionInfo.getProxyEventType();
+        if (eventType == ProxyEventType.BEFORE_QUERY) {
+            beforeQuery(executionInfo);
+        } else if (eventType == ProxyEventType.AFTER_QUERY) {
+            afterQuery(executionInfo);
+        }
+    }
```
Comment (gregturn) [inline src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java]: What if neither type is met?

File: src/test/java/io/r2dbc/proxy/callback/JdkProxyFactoryTest.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+import io.r2dbc.spi.Wrapped;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.lang.reflect.Proxy;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.mockito.Mockito.mock;
```
Comment (gregturn) [inline src/test/java/io/r2dbc/proxy/callback/JdkProxyFactoryTest.java]: Can we stick with one assertion toolkit?

File: src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+ */
+public class DefaultConnectionIdManager implements ConnectionIdManager {
+
+    private AtomicLong idCounter = new AtomicLong(0);
+
+    @Override
+    public String getId(Connection connection) {
+        String id = String.valueOf(this.idCounter.incrementAndGet());
+        return id;
+    }
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java]: Yes, definitely

File: src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+import io.r2dbc.spi.Statement;
+import io.r2dbc.spi.Wrapped;
+
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+
+/**
+ * ProxyFactory using JDK dynamic proxy.
+ *
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java]: Yes, will do

File: src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+
+    @Override
+    public Result createResult(Result result, QueryExecutionInfo queryExecutionInfo) {
+        return (Result) Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(),
+            new Class<?>[]{Result.class, Wrapped.class, ConnectionHolder.class},
+            new ResultInvocationHandler(result, queryExecutionInfo, this.proxyConfig));
+    }
+
+    public static class ConnectionFactoryInvocationHandler implements InvocationHandler {
+
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java]: No, they don't need to, I'll change this and others to private.

File: src/main/java/io/r2dbc/proxy/callback/ProxyConfig.java [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+    private ProxyFactory proxyFactory = new JdkProxyFactory();
+
+    {
+        this.proxyFactory.setProxyConfig(this);
+    }
+
+    //    public ProxyConfig(ProxyFactory proxyFactory) {
+    //        this.proxyFactory = proxyFactory;
+    //    }
+
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ProxyConfig.java]: Yup

File: src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java (hunk 1 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java]: Yes, the return value is the wrapped(proxied) input value.
Probably, rename each method to `createProxy...` for better naming??

Example usage would be:
```java
Connection proxiedConnecition = proxyFactory.createProxyConnection(originalConnection, connectionInfo);
```

File: src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java (hunk 2 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java]: Good Idea. I'll convert to Optional

File: src/main/java/io/r2dbc/proxy/callback/ReactiveBatchCallback.java (hunk 3 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ReactiveBatchCallback.java]: Um, yes, it can, but isn't it more common to initialize collection at declaration if they are not assigned by constructor?

File: src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java (hunk 4 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java]: Another way would be to compare `Method` object.

However, I think there are drawbacks for `Method` comparison.
- Need to programmatically prepare `Method` objects.
- Handling of overloading methods (same name but different args)

So, even though comparing with String method name ties to literal method name typing, it is much easier to use for controlling logic flow.

File: src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionFactoryCallback.java (hunk 5 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionFactoryCallback.java]: ok

File: src/main/java/io/r2dbc/proxy/listener/CompositeProxyExecutionListener.java (hunk 6 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/listener/CompositeProxyExecutionListener.java]: yes.

File: src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java (hunk 7 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java]: Hm, the calling logic kinds of ensures event type is either `BEFORE_METHOD` or `AFTER_METHOD`.

Actually, this is one of the part I changed the way from datasource-proxy jdbc to r2dbc.

In dsp-jdbc, instead of having this dispatcher method, proxy logic directly calls
`beforeMethod()` or `afterMethod()`. So, probably that might be better style than dispatching in default method like here.

I'll remove this method and directly call before/after methods.

File: src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java (hunk 8 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/listener/ProxyExecutionListener.java]: will remove this and call before/after directly

File: src/test/java/io/r2dbc/proxy/callback/JdkProxyFactoryTest.java (hunk 9 above)
Comment (ttddyy) [reply src/test/java/io/r2dbc/proxy/callback/JdkProxyFactoryTest.java]: OK, seems like  r2dbc projects uses assertj for all assertions.
I'll convert to it.

File: src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java [hunk 14]
```diff
@@ -0,0 +1,290 @@
...
+                executionInfo.setResult(result);
+                executionInfo.setThrown(thrown);
+
+                Instant currentTime = this.clock.instant();
+                Duration executionDuration = Duration.between(startTime, currentTime);
+                executionInfo.setExecuteDuration(executionDuration);
+
+                executionInfo.setProxyEventType(ProxyEventType.AFTER_METHOD);
+                listener.onMethodExecution(executionInfo);
+            }
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java]: Other ways I can think of is to create a StopWatch, or make a template for the pattern with Consumer.

Can you expand a little bit about try-resource object?

File: src/main/java/io/r2dbc/proxy/callback/ConnectionIdManager.java [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.proxy.callback;
+
+import io.r2dbc.spi.Connection;
+
+/**
+ * @author Tadaya Tsuyukubo
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ConnectionIdManager.java]: Yes, will do.

File: src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+package io.r2dbc.proxy.callback;
+
+import io.r2dbc.spi.Connection;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * Copy from datasource-proxy
+ *
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java]: Oops, missed this. will do.

File: src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java]: Essentially wrap things in a templated method. Perhaps Consumer or Supplier is a best fit. Try-Resource is something that must eventually be closed like a file, so maybe my comment was a little off.

File: src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (gregturn) [reply src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java]: Or consider package private. I’m a fan of default visibility.

File: src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java (hunk 1 above)
Comment (gregturn) [reply src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java]: Much better.

File: src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java (hunk 2 above)
Comment (gregturn) [reply src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java]: Check first. In some scenarios Optional adds a bit of overhead. They actually backed off doing it in Spring Data Commons.

File: src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java]: Tried to template the flow, but since it involves with Flux operations, not quite readable than directly putting logic on each Flux methods.
Instead, just made a simple `StopWatch` utility class, and delegated the `Duration` calculation, which seems cleaner approach.

File: src/main/java/io/r2dbc/proxy/callback/ConnectionIdManager.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ConnectionIdManager.java]: Added

File: src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/DefaultConnectionIdManager.java]: done

File: src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/JdkProxyFactory.java]: changed them to package private

File: src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java (hunk 1 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java]: Renamed to `createProxy...`

File: src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java (hunk 2 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java]: These method are utility for users and more for corner case use to retrieve original `Connection` object.
Not in the hot path. So, converted to `Optional`.

File: src/test/java/io/r2dbc/proxy/callback/JdkProxyFactoryTest.java (hunk 9 above)
Comment (ttddyy) [reply src/test/java/io/r2dbc/proxy/callback/JdkProxyFactoryTest.java]: Converted all jupiter `assert[Equals/Same/True/False/Throws]` to AssertJ assertions.
phew~ :)

**Comment by ttddyy** [comment]
@nebhale OK, got it. I'll update them.
Meanwhile, I just pushed one small commit to fix package directory structure.

File: src/main/java/io/r2dbc/proxy/callback/ProxyConfig.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+        this.proxyFactory.setProxyConfig(this);
+    }
+
+    public ProxyFactory getProxyFactory() {
+        return proxyFactory;
+    }
+
+    public void setProxyFactory(ProxyFactory proxyFactory) {
+        this.proxyFactory = proxyFactory;
+        this.proxyFactory.setProxyConfig(this);
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/callback/ProxyConfig.java]: That cycle is fishy. I understand that you need to create proxies and pass on some configuration. How about encapsulating either `ProxyFactory` with a `ProxyFactoryFactory` that then accepts a runtime configuration object? Not sure this is the best approach but right now I lack a better idea.

File: src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+/**
+ * Defines factory methods to create proxy of SPI classes.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public interface ProxyFactory {
+
+    void setProxyConfig(ProxyConfig proxyConfig);
+
+    ConnectionFactory createProxyConnectionFactory(ConnectionFactory connectionFactory);
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/callback/ProxyFactory.java]: How about reducing API surface in general by providing a single `<T> wrap(T delegate, Object... args)` method? The routing could be handled within the actual factory.

File: src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class ProxyUtils {
+
+    private ProxyUtils() {
+    }
+
+    @SuppressWarnings("unchecked")
+    public static Optional<Connection> getOriginalConnection(Connection connection) {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java]: How about sticking to `unwrap` terminology? Something like `unwrapConnection`?

File: src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java (hunk 2 above)
Comment (mp911de) [reply src/main/java/io/r2dbc/proxy/callback/ProxyUtils.java]: I also think non-hot code paths are fine for `Optional<…>` usage.

File: src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java (hunk 4 above)
Comment (mp911de) [reply src/main/java/io/r2dbc/proxy/callback/ReactiveConnectionCallback.java]: Using strings is quite a common pattern and the `Method` approach is significantly more difficult, as someone might redeclare a method and then it's a different one than retrieved from the interface.

File: src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java [hunk 20]
```diff
@@ -0,0 +1,284 @@
...
+import java.util.function.Consumer;
+
+import static java.util.stream.Collectors.toSet;
+
+/**
+ * Defines methods to augment execution of proxy methods used by child classes.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public abstract class CallbackSupport {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java]: How about naming types in this hierarchy `InvocationHandler` (this class would be named `InvocationHandlerSupport`)? Reflects better the actual purpose.

File: src/main/java/io/r2dbc/proxy/core/Binding.java [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+package io.r2dbc.proxy.core;
+
+/**
+ * Represent an operation of {@link io.r2dbc.spi.Statement#bind}.
+ *
+ * @author Tadaya Tsuyukubo
+ * @see Bindings.IndexBinding
+ * @see Bindings.IdentifierBinding
+ */
+public interface Binding {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/core/Binding.java]: Binding keeps reappearing across various implementations. That's just something I noticed, don't want to fix that right now.

File: src/main/java/io/r2dbc/proxy/core/BindingValue.java [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+package io.r2dbc.proxy.core;
+
+/**
+ * Represent a value for {@link io.r2dbc.spi.Statement#bind} and {@link io.r2dbc.spi.Statement#bindNull} operations.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public interface BindingValue {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/core/BindingValue.java]: How about `BoundValue`? Users don't have a chance to change it anyways.

File: src/main/java/io/r2dbc/proxy/core/BindingValue.java [hunk 23]
```diff
@@ -0,0 +1,61 @@
...
+
+    Object getValue();
+
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/core/BindingValue.java]: An `isNull()` method would be a good fit here.

File: src/main/java/io/r2dbc/proxy/core/ConnectionInfo.java [hunk 24]
```diff
@@ -0,0 +1,129 @@
...
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * {@link Connection} related information.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class ConnectionInfo {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/core/ConnectionInfo.java]: Let's turn that into an interface and make the default implementation package-private.

File: src/main/java/io/r2dbc/proxy/core/MethodExecutionInfo.java [hunk 25]
```diff
@@ -0,0 +1,148 @@
...
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Hold method execution related information.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class MethodExecutionInfo {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/core/MethodExecutionInfo.java]: Spring/Zipkin have various levels of exposure regarding this kind of object structures. E.g. `MethodExecutionInfo` (exposing read-only methods) and `MutableMethodExecutionInfo` so things get clear very quickly what the intent is.

File: src/main/java/io/r2dbc/proxy/core/QueryExecutionInfo.java [hunk 26]
```diff
@@ -0,0 +1,246 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Hold query execution related information.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class QueryExecutionInfo {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/core/QueryExecutionInfo.java]: See `MethodExecutionInfo` comment.

File: src/main/java/io/r2dbc/proxy/support/QueryExecutionInfoFormatter.java [hunk 27]
```diff
@@ -0,0 +1,434 @@
...
+import java.util.function.Function;
+
+import static java.util.stream.Collectors.joining;
+
+/**
+ * Convert {@link QueryExecutionInfo} to {@code String}.
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class QueryExecutionInfoFormatter implements Function<QueryExecutionInfo, String> {
```
Comment (mp911de) [inline src/main/java/io/r2dbc/proxy/support/QueryExecutionInfoFormatter.java]: That's a decent approach. Care to make `QueryExecutionInfoFormatter` immutable so it can be used similar to Java 8's Date/Time Formatters.

File: src/main/java/io/r2dbc/proxy/ProxyConnectionFactoryBuilder.java [hunk 28]
```diff
@@ -0,0 +1,133 @@
...
+    public ProxyConnectionFactoryBuilder(ConnectionFactory delegate) {
+        this.delegate = delegate;
+    }
+
+    public static ProxyConnectionFactoryBuilder create(ConnectionFactory delegate) {
+        Objects.requireNonNull(delegate, "ConnectionFactory to delegate is required");
+        return new ProxyConnectionFactoryBuilder(delegate);
+    }
+
+    public static ProxyConnectionFactoryBuilder create(ConnectionFactory delegate, ProxyConfig proxyConfig) {
```
[the remark this answers is not in the document]
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/ProxyConnectionFactoryBuilder.java]: So, something like this?

```java
public class ProxyConnectionFactory {
  public static ProxyConnectionFactoryBuilder create(ConnectionFactory original) {
    return new ProxyConnectionFactoryBuilder(original);
  }
}

public class ProxyConnectionFactoryBuilder {
  public ConnectionFactory build() {...}
  public ProxyConnectionFactoryBuilder listener(...) {...}
  ... other configuration methods ...
}
```

and, usage is:

```java
ConnectionFactory proxy = 
    ProxyConnectionFactory.create(actualConnectionFactory)
        .listener(...)
        .build();
```

compare to:

```java
// current implementation
ConnectionFactory proxy = 
    ProxyConnectionFactoryBuilder.create(actualConnectionFactory)
        .listener(...)
        .build();
```

File: src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java (hunk 20 above)
Comment (ttddyy) [reply src/main/java/io/r2dbc/proxy/callback/CallbackSupport.java]: I'm thinking to rename them to `CallbackHandler`.
So, 
- `CallbackSupport` => `CallbackHandlerSupport`
- `ReactiveStatementCallback` => `StatementCallbackHandler`
- `ReactiveConnectionFactoryCallback` => `ConnectionFactoryCallbackHandler`
- etc.

**Comment by ttddyy** [comment]
ok, I have updated the PR.

Summary of changes:
- Add more javadoc
- Add `Assert` class
  - Nullability check
- Restrict visibility of classes
- Use `wrap`, `unwrap` terminology
- Binding related
  - Rename `BindingValue` to `BoundValue`
  - Use `isNull` and consolidate `NullBindingValue` and `SimpleBindingValue`
- Convert `ConnectionInfo`, `MethodExecutionInfo`, `QueryExecutionInfo` to interface
- `ProxyFactoryFactory` to replace circular reference
- Make `Clock` configurable via `ProxyConfig`
- Make `[Query|Method]ExecutionInfoFormatter` immutable
- Rename `Reactive~Callback` to `~CallbackHandler`
- Refactor `JdkProxyFactory` with added `CallbackHandler` interface

----

Regarding visibility of the classes, I changed many of the classes to package scoped final classes. So, it is pretty much locked down and users cannot see or reuse most of them.

In datasource-proxy, I took the other approach. I made classes very open, so that user can reuse them either by overriding them or composing mix of user's custom classes and existing classes.

I just want to make sure in r2dbc projects, rule of thumbs is to hide as much as possible first.
Then, if user requests opening up a class and it is reasonable to do, then we may widen the scope.

**Comment by nebhale** [comment]
> I just want to make sure in r2dbc projects, rule of thumbs is to hide as much as possible first.
Then, if user requests opening up a class and it is reasonable to do, then we may widen the scope.

You're on the right path here.  We consider public APIs to be supported with compatibility guarantees which means the larger the surface area, the less future flexibility we have.  We strongly follow the open/closed principal as well in order to manage the area that bugs could be lurking due to ill-defined contracts.

File: src/test/java/io/r2dbc/proxy/support/MethodExecutionInfoFormatterTest.java [hunk 29]
```diff
@@ -0,0 +1,113 @@
...
+
+        // String#indexOf(int) method
+        Method method = ReflectionUtils.findMethod(String.class, "indexOf", int.class);
+
+        Long target = 100L;
+
+        ConnectionInfo connectionInfo = mock(ConnectionInfo.class);
+        when(connectionInfo.getConnectionId()).thenReturn("ABC");
+
+        MethodExecutionInfo executionInfo = mock(MethodExecutionInfo.class);
```
Comment (nebhale) [inline src/test/java/io/r2dbc/proxy/support/MethodExecutionInfoFormatterTest.java]: This isn't going to be a binding requirement, but I'd recommend creating a `test` package with full-fledged Mock types for all (or perhaps most important) interface types.  It's pretty obvious that both users _and_ the project need to create fake versions of these types and rather than requiring users to include a mocking library, and write the behavior, wouldn't it be nicer to say:

```java
MockMethodExecutionInfo executionInfo = MockMethodExecutionInfo.builder()
    .threadId(5L)
    .connectionInfo(MockConnectionInfo.builder()
        .connectionId("ABC")
        .build())
    .executeDuration(Duration.ofMillis(23))
    .method(String.class, "indexOf", int.class)
    .target(100L)
    .build()

File: src/test/java/io/r2dbc/proxy/support/QueryExecutionInfoFormatterTest.java [hunk 30]
```diff
@@ -0,0 +1,546 @@
...
+public class QueryExecutionInfoFormatterTest {
+
+    @Test
+    void batchExecution() {
+
+        ConnectionInfo connectionInfo = mock(ConnectionInfo.class);
+        when(connectionInfo.getConnectionId()).thenReturn("conn-id");
+
+        // Batch Query
+        QueryExecutionInfo execInfo = mock(QueryExecutionInfo.class);
```
Comment (nebhale) [inline src/test/java/io/r2dbc/proxy/support/QueryExecutionInfoFormatterTest.java]: This isn't going to be a binding requirement, but I'd recommend creating a `test` package with full-fledged Mock types for all (or perhaps most important) interface types.  It's pretty obvious that both users _and_ the project need to create fake versions of these types and rather than requiring users to include a mocking library, and write the behavior, wouldn't it be nicer to say:

```java
MockQueryExecutionInfo executionInfo = MockQueryExecutionInfoInfo.builder()
    .threadName("my-thread")
    .threadId(99L)
    .connectionInfo(MockConnectionInfo.builder()
        .connectionId("conn-id")
        .build())
    .success(true)
    .executeDuration(Duration.ofMillis(35))
    .type(ExecutionType.BATCH)
    .batchSize(20)
    .bindingsSize(10)
    .queries("SELECT A", "SELECT B")
    .build()

File: src/main/java/io/r2dbc/proxy/listener/LifeCycleListener.java [hunk 31]
```diff
@@ -0,0 +1,253 @@
...
+ * @author Tadaya Tsuyukubo
+ * @see LifeCycleExecutionListener
+ */
+public interface LifeCycleListener {
+
+    //
+    // for ConnectionFactory
+    //
+
+    default void beforeCreateOnConnectionFactory(MethodExecutionInfo methodExecutionInfo) {
```
Comment (nebhale) [inline src/main/java/io/r2dbc/proxy/listener/LifeCycleListener.java]: Need documentation on these methods as well, _especially_ around nullability guarantees for implementors.

File: src/main/java/io/r2dbc/proxy/core/QueryInfo.java [hunk 32]
```diff
@@ -0,0 +1,46 @@
...
+ *
+ * @author Tadaya Tsuyukubo
+ */
+public class QueryInfo {
+
+    private final String query;
+
+    private final List<Bindings> bindingsList = new ArrayList<>();
+
+    public QueryInfo(String query) {
```
Comment (nebhale) [inline src/main/java/io/r2dbc/proxy/core/QueryInfo.java]: Needs documentation.
